code/renderers/vue3/src/docs/sourceDecorator.ts

```diff
@@ -176,141 +187,133 @@ function getTemplates(renderFn: any): [] {
+            (type as any).__docgenInfo?.displayName;
+      const propsSource = generateAttributesSource(attributes, args, argTypes, byRef);
+      return childSources.trim() === ''
+        ? `<${name} ${propsSource}/>`
+        : `<${name} ${propsSource}>${childSources}</${name}>`;
     }
 
-    return `<${name} ${props}/>`;
+    return null;
   };
-  // get one component or multiple
-  const components = isArray(compOrComps) ? compOrComps : [compOrComps];
-
-  const source = Object.keys(components)
-    .map((key: any) => `${generateComponentSource(components[key])}`)
-    .join(`\n`);
-  return source;
-}
 
-/**
- * create Named Slots content in source
- * @param slotProps
- * @param slotArgs
- */
-
-function createNamedSlots(slotArgs: ArgEntries, slotProps: ArgEntries, byRef?: boolean) {
-  if (!slotArgs) return '';
-  const many = slotProps.length > 1;
-  return slotArgs
-    .map(([key, value]) => {
-      const content = !byRef ? JSON.stringify(value) : `{{ ${key} }}`;
-      return many ? `  <template #${key}>${content}</template>` : `  ${content}`;
-    })
-    .join('\n');
-}
-
-function getArgsInAttrs(args: Args, attributes: Attribute[]) {
-  return Object.keys(args).reduce((acc, prop) => {
-    if (attributes?.find((attr: any) => attr.name === 'v-bind')) {
-      acc[prop] = args[prop];
-    }
-    const attribute = attributes?.find(
-      (attr: any) => attr.name === prop || attr.name === `:${prop}`
-    );
-    if (attribute) {
-      acc[prop] = attribute.name === `:${prop}` ? args[prop] : attribute.value;
-    }
-    if (Object.keys(acc).length === 0) {
-      attributes?.forEach((attr: any) => {
-        acc[attr.name] = JSON.parse(JSON.stringify(attr.value));
-      });
-    }
-    return acc;
-  }, {} as Record<string, any>);
+  const componentsOrNodes = Array.isArray(componentOrNodes) ? componentOrNodes : [componentOrNodes];
+  const source = componentsOrNodes
+    .map((componentOrNode) => generateComponentSource(componentOrNode))
+    .join(' ');
+  return source || null;
 }
 
-/**
- * format prettier for vue
- * @param source
- */
-
 /**
  *  source decorator.
  * @param storyFn Fn
  * @param context  StoryContext
  */
 export const sourceDecorator = (storyFn: any, context: StoryContext<Renderer>) => {
-  const channel = addons.getChannel();
   const skip = skipSourceRender(context);
   const story = storyFn();
 
-  let source: string;
-
-  useEffect(() => {
-    if (!skip && source) {
-      const { id, unmappedArgs } = context;
-      channel.emit(SNIPPET_RENDERED, { id, args: unmappedArgs, source, format: 'vue' });
-    }
-  });
-
-  if (skip) {
-    return story;
-  }
-
-  const { args = {}, component: ctxtComponent, argTypes = {} } = context || {};
-  const components = getTemplates(context?.originalStoryFn);
+  watch(
+    () => context.args,
+    () => {
+      if (!skip) {
+        generateSource(context);
+      }
+    },
+    { immediate: true, deep: true }
+  );
+  return story;
```

Comment:
Shouldn't we do something like:
```ts
() => {
  const story = storyFn();
  watch(
    () => context.args,
    () => {
      if (!skip) {
        generateSource(context);
      }
    },
    { immediate: true, deep: true }
  );
  return story;
};
```

I mean returning a functional component, instead of directly calling the storyFn?